src/StmtToViz.cpp

```diff
@@ -0,0 +1,2900 @@
+    std::map<uint64_t, std::regex> _markers;
+
+    // Mapping of IR nodes to their asm line numbers
+    std::map<uint64_t, int> _lnos;
+
+    using IRVisitor::visit;
+
+    void visit(const ProducerConsumer *op) override {
+        // Generate asm marker
+        _markers[(uint64_t)op] = gen_prodcons_asm_marker(gen_prodcons_id(), op->name, op->is_producer);
+        // Continue traversal
+        IRVisitor::visit(op);
+    }
+
+    void visit(const For *op) override {
+        // Generate asm marker
+        _markers[(uint64_t)op] = gen_loop_asm_marker(gen_loop_id(), op->name);
+        // Continue traversal
+        IRVisitor::visit(op);
+    }
+};
+
+/******************* HTMLCodePrinter Class *******************/
+// Prints IR code in HTML. Very similar to generating a stmt
+// file, except that the generated html is more interactive.
+template<typename T>
+class HTMLCodePrinter : public IRVisitor {
+public:
+    HTMLCodePrinter(T &os)
+        : stream(os), _id(0), context_stack(1, 0) {
+    }
+
+    void init_cost_info(IRCostModel cost_model) {
+        _cost_model = std::move(cost_model);
+    }
+
+    void print(const Module &m) {
+        // Generate a unique ID for this module
+        int id = gen_unique_id();
+
+        // Enter new scope for this module
+        scope.push(m.name(), id);
+
+        // The implementation currently does not support submodules.
+        // TODO: test it out on a benchmark with submodules
+        internal_assert(m.submodules().empty()) << "StmtToViz does not yet support submodules.";
+
+        // Open div to hold this module
+        print_opening_tag("div", "Module");
+
+        // Generate the show hide icon/text buttons
+        print_toggle_anchor_opening_tag(id);
+
+        // -- print icon
+        print_show_hide_icon(id);
+
+        // -- print text
+        print_opening_tag("span", "matched");
+        print_html_element("span", "keyword", "module");
+        print_text(" name=" + m.name() + ", target=" + m.target().to_string());
+        print_closing_tag("span");
+
+        print_toggle_anchor_closing_tag();
+
+        // Open code block to hold module body
+        print_html_element("span", "matched", " {");
+
+        // Open indented div to hold body code
+        print_opening_tag("div", "indent ModuleBody", id);
+
+        // Print module buffers
+        for (const auto &buf : m.buffers()) {
+            print(buf);
+        }
+
+        // Print module functions
+        for (const auto &fn : m.functions()) {
+            print(fn);
+        }
+
+        // Close indented div holding body code
+        print_closing_tag("div");
+
+        // Close code block holding module body
+        print_html_element("span", "matched ClosingBrace cb-" + std::to_string(id), "}");
+
+        // Close div holding this module
+        print_closing_tag("div");
+
+        // Pop out to outer scope
+        scope.pop(m.name());
+    }
+
+private:
+    // Handle to output file stream
+    T &stream;
```

Comment:
Since we're holding a mutable reference to an external thing here, probably wise to make this class non-copyable and non-movable to be sure the reference can't go stale